### PR TITLE
Include certificates in LDS response

### PR DESCRIPTION
**Pull request by jerith**

### CODE REVIEW

File: lib/relay/certs.ex [hunk 1]
```diff
@@ -18,15 +31,42 @@ defmodule Relay.Certs do
...
+  to support self-signed certs (which may look a lot like CA certs), we assume
+  that if there's only one cert in the PEM data it's the one we want.
+  """
+  @spec get_end_entity_hostnames(binary | [pem_entry]) :: [String.t]
+  def get_end_entity_hostnames(pem_data) do
+    get_certs(pem_data)
+    |> get_end_entity_certs()
     |> Enum.flat_map(&get_hostnames/1)
-    |> Enum.uniq()
+    |> Enum.uniq
```
[the remark this answers is not in the document]
Comment (jerith) [reply lib/relay/certs.ex]: Oops. Thanks.

File: test/relay/certs_test.exs [hunk 2]
```diff
@@ -3,21 +3,67 @@ defmodule Relay.CertsTest do
...
+  defp read_cert!(certfile) do
+    Path.join("../support/", certfile)
+    |> Path.expand(__DIR__)
+    |> File.read!()
+  end
+
+  test "pem encode" do
+    [key, cert, cacert] = :public_key.pem_decode(read_cert!("localhost.pem"))
+    pem_data = :public_key.pem_encode([key, cert, cacert])
+    assert ^pem_data = Certs.pem_encode(pem_data)
```
[the remark this answers is not in the document]
Comment (jerith) [reply test/relay/certs_test.exs]: Yeah, good point.

File: test/relay/certs_test.exs [hunk 3]
```diff
@@ -3,21 +3,67 @@ defmodule Relay.CertsTest do
...
+    assert ["demo-nosan.example.com"] = Enum.sort(hostnames)
+  end
+
+  test "collect hostnames for self-signed cert with SAN" do
+    hostnames = Certs.get_end_entity_hostnames(read_cert!("demo.pem"))
+    assert ["demo.example.com", "demo.example.net"] = Enum.sort(hostnames)
+  end
+
+  test ":relay_pk_utils.get_end_entity_certs" do
+    # This tests an edge case that isn't exercised through Relay.Certs.
```
[the remark this answers is not in the document]
Comment (jerith) [reply test/relay/certs_test.exs]: Definitely, thanks.

File: lib/relay/certs.ex [hunk 4]
```diff
@@ -18,15 +31,42 @@ defmodule Relay.Certs do
...
-  Extracts hostnames from all end-entity certs in the given PEM data.
+  Returns a list of all certificates in the given PEM data.
   """
-  @spec get_end_entity_hostnames(binary | [:public_key.pem_entry]) :: [String.t]
-  def get_end_entity_hostnames(pem_data) when is_binary(pem_data),
-    do: get_end_entity_hostnames(:public_key.pem_decode(pem_data))
+  @spec get_certs(binary | [pem_entry]) :: [pem_entry]
+  def get_certs(pem_data) do
+    pem_decode(pem_data)
+    |> Enum.filter(fn {:Certificate, _, _} -> true; _ -> false end)
```
[the remark this answers is not in the document]
Comment (jerith) [reply lib/relay/certs.ex]: It lets you put multiple things (like lambda clauses) on a single line. I've seldom seen it used except for things like this.

File: lib/relay/demo/certs.ex [hunk 5]
```diff
@@ -144,29 +144,37 @@ defmodule Relay.Demo.Certs do
...
-      filters: [default_http_conn_manager_filter(name)]
+      filter_chain_match: FilterChainMatch.new(sni_domains: sni_domains),
+      filters: [default_http_conn_manager_filter(name)],
+      tls_context: tls_context
     )
   end
 
+  defp inline_pem(pem_data) do
+    alias Envoy.Api.V2.Core.DataSource
+    DataSource.new(specifier: {:inline_string, Certs.pem_encode(pem_data)})
```
Comment (JayH5) [inline lib/relay/demo/certs.ex]: Now that we're using Protobufs, I'm not sure `inline_bytes` is any different to `inline_string` 🤔 (When we were using JSON, the bytes encoding was a base64-encoded string..which didn't make sense to use since PEM is already a base64-encoded format.)

Anyway, just a thought mostly.
